src/merkle/merkle_tree.rs

```diff
@@ -0,0 +1,278 @@
+use std::hash::{DefaultHasher, Hash, Hasher};
+
+pub enum SiblingsHash {
+    LeftSibling(u64),
+    RightSibling(u64),
+}
+
+#[derive(Clone)]
+struct MerkleNode {
+    hash_value: u64,
+    left_son: Option<Box<MerkleNode>>,
+    right_son: Option<Box<MerkleNode>>,
+}
+pub struct MerkleTree<H: Hash + Clone> {
+    merkle_root: MerkleNode,
+    leafs: Vec<H>,
+}
+
+impl MerkleNode {
+    pub fn new(
+        hash_value: u64,
+        left_son: Option<Box<MerkleNode>>,
+        right_son: Option<Box<MerkleNode>>,
+    ) -> Self {
+        Self {
+            hash_value,
+            left_son,
+            right_son,
+        }
+    }
+}
+
+impl<H: Hash + Clone> MerkleTree<H> {
+    pub fn new(transactions: Vec<H>) -> Self {
+        Self::create_tree(transactions)
+    }
+
+    fn create_parent_from_siblings(
+        left_son: Option<Box<MerkleNode>>,
+        mut right_son: Option<Box<MerkleNode>>,
+    ) -> MerkleNode {
+        let mut hasher = DefaultHasher::new();
+
+        if let Some(left_sibling) = &left_son {
+            left_sibling.hash_value.hash(&mut hasher);
+            if let Some(right_sibling) = &right_son {
+                right_sibling.hash_value.hash(&mut hasher);
+            } else {
+                right_son = left_son.clone();
+                left_sibling.hash_value.hash(&mut hasher);
+            }
+        }
+        MerkleNode::new(hasher.finish(), left_son, right_son)
+    }
+
+    fn create_tree(transactions: Vec<H>) -> MerkleTree<H> {
+        if transactions.is_empty() {
+            return Self {
+                merkle_root: MerkleNode::new(0, None, None),
+                leafs: transactions,
+            };
+        }
```

Comment:
It might be more appropriate to return an `Err` in this case, as having a tree with no elements could lead to unexpected situations.